EddieMachete/marketing-parameter-forwarder
# Set Marketing Cookie

### User story:
As a marketing specialist, when deploying a campaign,
I want our public website to forward our UTM and other marketing parameters when exiting the site
so that we can piece together the visitor's journey and track the success of our campaigns.

### Business logic:
1. When a page on the site loads, the system checks if we are coming from an external source
   1. If we are coming from an external page, the system updates the marketing cookies
       1. The system retrieves the white listed parameters from the query string
       2. The system ensures parameters have been cleaned up to protect against script injections
       3. The system sets the marketing cookies
       4. The system retrieves the marketing parameters from the URL
   2. If we are coming from an internal page
       1. The system retrieves the marketing parameters from the existing cookie
2. The system updates the marketing links on the page by appending the marketing parameters to the search string

## Testing

1. In the terminal, execute the command `npm start`
2. On the browser, navigate to http://localhost:44333/?utm_source=google&utm_medium=cpc&utm_campaign=spring_sale&utm_term=running_shoes&utm_content=textlink
3. Confirm that the `external page` link has the UTM parameters specified in the url
4. Navigate to an internal page and confirm that the `external page` link retains the UTM parameters
5. Navigate to the external page and confirm the URL in the navigation bar contains the UTM parameters originally used
